### PR TITLE
Deployment preparation: `ChainlinkRateProviderFactory`

### PR DESCRIPTION
**Pull request by baileyspraggins**
# Deployment preparation: `ChainlinkRateProviderFactory`

Build-info: https://github.com/balancer/metastable-rate-providers/commit/8793212b2567b2df81d6162bcb125067a5cd4ba1

## Deployment preparation tasks checklist:

- [x] Deployments `README` is updated with the new task name and link
- [x] Task `README` has a description and links to the artifacts
- [x] `build-info` is updated with artifacts compiled at the specified commit
- [x] Artifacts are generated from `build-info`
- [x] New task has a working fork test 

## Code checklist:

- [x] The diff is legible and has no extraneous changes 
- [x] Complex code has been commented in tests and scripts
- [x] Fork test block number and test cases are appropriate for this deployment
- [x] The base branch is either `master`, or there's a description of how to merge

## Issue Resolution

N/A

### CODE REVIEW

File: tasks/20230717-chainlink-rate-provider-factory/test/test.fork.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    const tx = await (await rateProviderFactory.create(usdcPriceFeed.address)).wait();
+    const event = await expectEvent.inReceipt(tx, 'RateProviderCreated');
+
+    rateProvider = await ethers.getContractAt(chainlinkRateProviderArtifact.abi, event.args.rateProvider);
+
+    expect(rateProvider.address).to.not.equal(ZERO_ADDRESS);
+    expect(await rateProviderFactory.isRateProviderFromFactory(rateProvider.address)).to.be.true;
+  });
+
+  it('get rate', async () => {
```
Comment (rabmarut) [inline tasks/20230717-chainlink-rate-provider-factory/test/test.fork.ts:35]: I think `it` descriptions should explain the goal of the test.
```suggestion
  it('rate is about 1 USD per USDC', async () => {
```

File: tasks/20230717-chainlink-rate-provider-factory/test/test.fork.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+  before('run task', async () => {
+    task = new Task('20230717-chainlink-rate-provider-factory', TaskMode.TEST, getForkedNetwork(hre));
+    await task.run({ force: true });
+
+    usdcPriceFeed = await ethers.getContractAt(aggregatorV3InterfaceABI, usdcPriceFeedAddress);
+    rateProviderFactory = await task.deployedInstance('ChainlinkRateProviderFactory');
+  });
+
+  it('create a ChainLinkRateProvider', async () => {
```
Comment (rabmarut) [inline tasks/20230717-chainlink-rate-provider-factory/test/test.fork.ts:25]: Creating the rate provider can be a `before` operation whereas testing specific pieces of functionality should be broken out into `it`. It's okay for the `before` operation to contain some small tests, such as verifying the event and contract registration.
```suggestion
  before('create a ChainLinkRateProvider', async () => {
```

**Review by rabmarut — CHANGES_REQUESTED** [review]

**Review by rabmarut — APPROVED** [review]

File: tasks/20230717-chainlink-rate-provider-factory/artifact/AggregatorV3Interface.json
Comment (jubeira) [inline tasks/20230717-chainlink-rate-provider-factory/artifact/AggregatorV3Interface.json:1]: This file can be removed (see comments in test below)

File: tasks/20230717-chainlink-rate-provider-factory/artifact/ChainlinkRateProvider.json
Comment (jubeira) [inline tasks/20230717-chainlink-rate-provider-factory/artifact/ChainlinkRateProvider.json:1]: This file can be removed (see comments in test below)

File: tasks/20230717-chainlink-rate-provider-factory/readme.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# 2023-07-17 - Chainlink Rate Provider Factory
+
+Deployment of `ChainlinkRateProviderFactory`, which creates and deploys `ChainlinkRateProvider` contracts.
+
+## Useful Files
+
+- [`AggregatorV3Interface` artifact](./artifact/AggregatorV3Interface.json)
+- [`ChainlinkRateProvider` artifact](./artifact/ChainlinkRateProvider.json)
```
Comment (jubeira) [inline tasks/20230717-chainlink-rate-provider-factory/readme.md:8]:
```suggestion
```

File: tasks/20230717-chainlink-rate-provider-factory/test/test.fork.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import hre, { ethers } from 'hardhat';
```
Comment (jubeira) [inline tasks/20230717-chainlink-rate-provider-factory/test/test.fork.ts:1]:
```suggestion
import hre from 'hardhat';
```

**Review by jubeira — COMMENTED** [review]
Looks pretty good!
I've verified that the build info matches with the one generated by https://github.com/balancer/metastable-rate-providers/commit/8793212b2567b2df81d6162bcb125067a5cd4ba1.

I've left a bunch of comments to simplify the test a bit, also matching the style of the rest of the tests. Other than that, it's good to go; great work!

File: tasks/20230717-chainlink-rate-provider-factory/artifact/AggregatorV3Interface.json
Comment (jubeira) [reply tasks/20230717-chainlink-rate-provider-factory/artifact/AggregatorV3Interface.json:1]: In short, these artifacts are not part of the contracts to be deployed and are used in the test only.

We have utilities that let you get a `Contract` instance in the tests by referencing it with the contract name, as long as it's included in the build info file (which is the case here).

**Comment by baileyspraggins** [comment]
> Looks pretty good! I've verified that the build info matches with the one generated by [balancer/metastable-rate-providers@8793212](https://github.com/balancer/metastable-rate-providers/commit/8793212b2567b2df81d6162bcb125067a5cd4ba1).
> 
> I've left a bunch of comments to simplify the test a bit, also matching the style of the rest of the tests. Other than that, it's good to go; great work!

@jubeira Thanks for mentioning the tools to get the contract instance using its name. Ive made the suggested changes. Should be good to go now.

**Pull request merged**